src/components/footer/RPCToggle.tsx

```diff
@@ -0,0 +1,196 @@
+import React, { FC, useCallback, useMemo, useState } from 'react'
+import {
+  Button,
+  Dialog,
+  DialogBody,
+  DialogCloseDefault,
+  DialogContent,
+  DialogOverlay,
+  DialogTrigger,
+  Icon,
+  IconTooltip,
+  Input,
+  InputElementRight,
+  InputGroup,
+  Popover,
+  PopoverContent,
+  PopoverTrigger
+} from 'gfx-component-lib'
+import { FooterItem, FooterItemContent, FooterItemProps } from '@/components/footer/FooterItem'
+import { EndPointName, useConnectionConfig, useDarkMode, USER_CACHE } from '@/context'
+import { Circle } from '@/components/common/Circle'
+import RadioOptionGroup from '@/components/common/RadioOptionGroup'
+import { testRPC } from '@/utils/requests'
+import useBreakPoint from '@/hooks/useBreakPoint'
+import useBoolean from '@/hooks/useBoolean'
+type RPCToggleProps = Omit<FooterItemProps, 'title'>
+const RPCToggle: FC<RPCToggleProps> = ({ ...rest }) => {
+  const { mode } = useDarkMode()
+  const { isMobile } = useBreakPoint()
+  const { endpointName, setEndpointName, latency } = useConnectionConfig()
+  const [RPC, setRPC] = useState<EndPointName>(endpointName)
+  const [rpcUrl, setRpcUrl] = useState('')
+  const [error, setError] = useState('')
+  const [isOpen, setIsOpen] = useBoolean(false)
+  const providerSrc = useMemo(
+    () => `/img/mainnav/provider_${endpointName.toLowerCase()}_${mode}.svg`,
+    [endpointName, mode]
+  )
+  const handleSave = useCallback(() => {
+    if (RPC === 'Custom') {
+      testRPC(rpcUrl).then((res) => {
+        if (res) {
+          const existingCache = JSON.parse(window.localStorage.getItem(USER_CACHE))
+          window.localStorage.setItem(USER_CACHE, JSON.stringify({ ...existingCache, endpoint: rpcUrl }))
+          setEndpointName(RPC)
+          setError('')
+          setIsOpen.off()
+        } else {
+          setError('Invalid RPC URL')
+        }
+      })
+
+      return
+    }
+    setError('')
```

Comment:
Here we have race.  setError, setIsOpen, setEndpointName will trigger twice.

Use await for more imperative way

Reply:
There is on race condition, return is within the if, not the promise.

Reply:
ah ok, now I see!